api/pull.go

```diff
@@ -0,0 +1,380 @@
+package api
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"github.com/ethereum/go-ethereum/common"
+	gethTypes "github.com/ethereum/go-ethereum/core/types"
+	"github.com/ethereum/go-ethereum/eth/filters"
+	"github.com/ethereum/go-ethereum/rpc"
+	"github.com/onflow/flow-evm-gateway/config"
+	"github.com/onflow/flow-evm-gateway/models"
+	"github.com/onflow/flow-evm-gateway/services/logs"
+	"github.com/onflow/flow-evm-gateway/storage"
+	errs "github.com/onflow/flow-evm-gateway/storage/errors"
+	"github.com/rs/zerolog"
+	"math/big"
+	"sync"
+	"time"
+)
+
+// filter defines a general resource filter that is used when pulling new data.
+//
+// Filters work by remembering the last height at which the data was pulled
+// and allow on next request to fetch all data that is missing.
+// Each filter is identified by a unique ID.
+type filter interface {
+	id() rpc.ID
+	nextHeight() uint64
+	updateUsed(height uint64)
+	expired() bool
+}
+
+// baseFilter is a base implementation that keeps track of
+// last height and filter unique ID.
+type baseFilter struct {
+	last   uint64
+	rpcID  rpc.ID
```

Comment:
I am guessing that only one client has the generated `rpcID`, so there's no need for adding any sync mechanism to the `updatedUsed` and `nextHeight` methods, right?

Reply:
It might be conflicting in a rare case, good point, I added write mutex on it